File: engine/predictedAeProfile.ts
// Predicted adverse-effect profile for the to-drug.
//
// Cross-references our existing ADVERSE_EFFECTS table (engine/adverseEffects.ts)
// and the to-drug's per-AE risk fields (sedation, epsRisk, prolactinRisk,
// qtcRisk, metabolicRisk, discontinuationSyndromeRisk) to surface a
// prioritized list of "watch out for these on the new drug".
//
// Two data sources, in priority order:
//   1. Per-drug risk fields on the JSON (qualitative 'low' | 'moderate'
//      | 'high' | 'very high'). Maps directly to a likelihood tier.
//   2. The reverse-lookup table in adverseEffects.ts (drug appears in
//      `causedBy` → likely; drug appears in `switchCandidates` → unlikely
//      relative to the from-drug).
//
// We deliberately don't try to give a numeric percentage. Cipriani 2018
// and Leucht 2013 NMAs report drug-vs-placebo odds ratios, not absolute
// risks, and copying those out of context would be misleading. Instead
// we surface tiers — clinicians know what "high prolactin risk" looks
// like for risperidone, they don't need a percentage.

import { ADVERSE_EFFECTS, type AdverseEffect } from './adverseEffects';
import type { Drug } from './types';

export type AeLikelihood = 'high' | 'moderate' | 'low' | 'lower-than-current' | 'unknown';

export interface PredictedAe {
  /** Reference to the adverse-effect entry (id, label, summary, etc.). */
  ae: AdverseEffect;
  likelihood: AeLikelihood;
  /** Why we chose this tier — for the "?" tooltip / breakdown row. */
  reason: string;
}

export interface PredictedAeProfile {
  toDrug: Drug;
  /** Sorted list — high → moderate → low → lower-than-current → unknown. */
  predictions: PredictedAe[];
}

const RISK_TO_LIKELIHOOD: Record<string, AeLikelihood> = {
  'very high': 'high',
  high: 'high',
  moderate: 'moderate',
  low: 'low',
};

/**
 * Generate a predicted AE profile for the to-drug, optionally
 * comparing against the from-drug to highlight what's *better* on the
 * new agent.
 */
export function predictAeProfile(toDrug: Drug, fromDrug?: Drug): PredictedAeProfile {
  const predictions: PredictedAe[] = [];

  for (const ae of ADVERSE_EFFECTS) {
    const pred = predictOne(ae, toDrug, fromDrug);
    if (pred) predictions.push(pred);
  }

  predictions.sort((a, b) => likelihoodRank(b.likelihood) - likelihoodRank(a.likelihood));

  return { toDrug, predictions };
}

function predictOne(
  ae: AdverseEffect,
  toDrug: Drug,
  fromDrug?: Drug,
): PredictedAe | null {
  // Two signal sources. We compute BOTH then take the higher tier so
  // that, e.g., aripiprazole shows high akathisia risk (reverse-lookup
  // says yes — it's the hallmark dose-limiting AE) even though its
  // overall epsRisk field reads 'low' (which is correct for the
  // parkinsonism subtype but misleading on its own for akathisia).

  // Source 1 — per-drug typed risk fields. Graded but coarse-bucketed.
  const fieldRisk = pickRiskField(toDrug, ae.id);
  const fieldTier: AeLikelihood | null = fieldRisk ? (RISK_TO_LIKELIHOOD[fieldRisk] ?? null) : null;

  // Source 2 — ADVERSE_EFFECTS reverse-lookup. Binary but specific.
  const inCauses = ae.causedBy.includes(toDrug.id);
  const inAvoids = ae.switchCandidates.includes(toDrug.id);
  const fromCauses = fromDrug ? ae.causedBy.includes(fromDrug.id) : false;

  let lookupTier: AeLikelihood | null = null;
  let lookupReason = '';
  if (inCauses && inAvoids) {
    // Edge case — shouldn't happen with curated data, but defensive.
    lookupTier = 'moderate';
    lookupReason = 'Listed in both candidates and causes.';
  } else if (inCauses) {
    lookupTier = 'high';
    lookupReason = 'Commonly caused by this drug.';
  } else if (inAvoids && fromCauses) {
    lookupTier = 'lower-than-current';
    lookupReason = `Recommended switch target for ${ae.label.toLowerCase()}.`;
  } else if (inAvoids) {
    lookupTier = 'low';
    lookupReason = 'Not typically caused by this drug.';
  }

  // Comparative hint always wins — it's the most useful signal in
  // a switching context.
  if (lookupTier === 'lower-than-current') {
    return { ae, likelihood: 'lower-than-current', reason: lookupReason };
  }

  if (!fieldTier && !lookupTier) return null;

  // Pick the higher-severity tier between the two signals.
  const winner: AeLikelihood = takeHigher(fieldTier, lookupTier);
  const reason =
    fieldTier && lookupTier
      ? winner === fieldTier
        ? `Drug profile: ${fieldRisk} risk.`
        : lookupReason
      : fieldTier
        ? `Drug profile: ${fieldRisk} risk.`
        : lookupReason;
  return { ae, likelihood: winner, reason };
}

function takeHigher(a: AeLikelihood | null, b: AeLikelihood | null): AeLikelihood {
  if (!a && !b) return 'unknown';
  if (!a) return b!;
  if (!b) return a;
  return likelihoodRank(a) >= likelihoodRank(b) ? a : b;
}

/**
 * Map an AE id to the matching risk field on the Drug object. We only
 * surface a few common AEs through the typed fields; everything else
 * falls through to the reverse-lookup table.
 */
function pickRiskField(drug: Drug, aeId: string): string | null {
  switch (aeId) {
    case 'sedation':
      return drug.sedation ?? null;
    case 'eps_akathisia':
      return drug.epsRisk ?? null;
    case 'hyperprolactinaemia':
      return drug.prolactinRisk ?? null;
    case 'qtc_prolongation':
      return drug.qtcRisk ?? null;
    case 'weight_gain':
      return drug.metabolicRisk?.score ?? null;
    case 'discontinuation_difficult':
      return drug.discontinuationSyndromeRisk?.score ?? null;
    default:
      return null;
  }
}

function likelihoodRank(l: AeLikelihood): number {
  switch (l) {
    case 'high': return 4;
    case 'moderate': return 3;
    case 'low': return 2;
    case 'lower-than-current': return 1;
    case 'unknown': return 0;
  }
}

export function likelihoodLabel(l: AeLikelihood): string {
  switch (l) {
    case 'high': return 'High likelihood';
    case 'moderate': return 'Moderate likelihood';
    case 'low': return 'Low likelihood';
    case 'lower-than-current': return 'Lower than current drug';
    case 'unknown': return 'Unknown';
  }
}

export function likelihoodColor(l: AeLikelihood): { bg: string; text: string } {
  switch (l) {
    case 'high':                 return { bg: 'bg-danger/15',  text: 'text-danger' };
    case 'moderate':             return { bg: 'bg-warning/15', text: 'text-warning' };
    case 'low':                  return { bg: 'bg-accent/15',  text: 'text-accent' };
    case 'lower-than-current':   return { bg: 'bg-to/15',      text: 'text-to' };
    case 'unknown':              return { bg: 'bg-border',     text: 'text-muted' };
  }
}

File: psychswitch_engine/lib/src/predicted_ae_profile.dart
// Predicted adverse-effect profile for the to-drug.
//
// Cross-references the ADVERSE_EFFECTS table (engine/adverse_effects.dart)
// and the to-drug's per-AE risk fields (sedation, epsRisk, prolactinRisk,
// qtcRisk, metabolicRisk, discontinuationSyndromeRisk) to surface a
// prioritized list of "watch out for these on the new drug".
//
// Two data sources, in priority order:
//   1. Per-drug risk fields on the JSON (qualitative 'low' | 'moderate'
//      | 'high' | 'very high'). Maps directly to a likelihood tier.
//   2. The reverse-lookup table in adverse_effects.dart (drug appears in
//      `causedBy` → likely; drug appears in `switchCandidates` →
//      unlikely relative to the from-drug).
//
// We deliberately don't try to give a numeric percentage. Cipriani 2018
// and Leucht 2013 NMAs report drug-vs-placebo odds ratios, not absolute
// risks, and copying those out of context would be misleading. Instead
// we surface tiers — clinicians know what "high prolactin risk" looks
// like for risperidone, they don't need a percentage.
//
// Dart port of engine/predictedAeProfile.ts.

import 'package:psychswitch_engine/adverse_effects.dart';
import 'package:psychswitch_engine/types/drug.dart';
import 'package:psychswitch_engine/types/enums.dart';

/// Likelihood tier for a single AE on the to-drug.
enum AeLikelihood {
  high('high'),
  moderate('moderate'),
  low('low'),
  lowerThanCurrent('lower-than-current'),
  unknown('unknown');

  const AeLikelihood(this.jsonValue);

  final String jsonValue;

  static AeLikelihood fromJson(String value) {
    for (final l in AeLikelihood.values) {
      if (l.jsonValue == value) return l;
    }
    throw ArgumentError.value(value, 'value', 'unknown AeLikelihood');
  }
}

/// One predicted AE row.
class PredictedAe {
  const PredictedAe({
    required this.ae,
    required this.likelihood,
    required this.reason,
  });

  /// Reference to the adverse-effect entry (id, label, summary, etc.).
  final AdverseEffect ae;

  final AeLikelihood likelihood;

  /// Why we chose this tier — for the "?" tooltip / breakdown row.
  final String reason;

  Map<String, dynamic> toJson() => <String, dynamic>{
        'ae': ae.toJson(),
        'likelihood': likelihood.jsonValue,
        'reason': reason,
      };
}

/// Full prediction output for a to-drug.
class PredictedAeProfile {
  const PredictedAeProfile({required this.toDrug, required this.predictions});

  final Drug toDrug;

  /// Sorted list — high → moderate → low → lower-than-current → unknown.
  final List<PredictedAe> predictions;
}

const Map<RiskLevel, AeLikelihood> _riskToLikelihood = <RiskLevel, AeLikelihood>{
  RiskLevel.veryHigh: AeLikelihood.high,
  RiskLevel.high: AeLikelihood.high,
  RiskLevel.moderate: AeLikelihood.moderate,
  RiskLevel.lowModerate: AeLikelihood.moderate,
  RiskLevel.low: AeLikelihood.low,
};

/// Generate a predicted AE profile for the to-drug, optionally
/// comparing against the from-drug to highlight what's *better* on the
/// new agent.
PredictedAeProfile predictAeProfile(Drug toDrug, [Drug? fromDrug]) {
  final predictions = <PredictedAe>[];
  for (final ae in adverseEffects) {
    final pred = _predictOne(ae, toDrug, fromDrug);
    if (pred != null) predictions.add(pred);
  }
  predictions.sort(
    (a, b) => _likelihoodRank(b.likelihood) - _likelihoodRank(a.likelihood),
  );
  return PredictedAeProfile(toDrug: toDrug, predictions: predictions);
}

PredictedAe? _predictOne(
  AdverseEffect ae,
  Drug toDrug,
  Drug? fromDrug,
) {
  // Source 1 — per-drug typed risk fields.
  final fieldRisk = _pickRiskField(toDrug, ae.id);
  final fieldTier = fieldRisk != null ? _riskToLikelihood[fieldRisk] : null;

  // Source 2 — adverseEffects reverse-lookup.
  final inCauses = ae.causedBy.contains(toDrug.id);
  final inAvoids = ae.switchCandidates.contains(toDrug.id);
  final fromCauses =
      fromDrug != null && ae.causedBy.contains(fromDrug.id);

  AeLikelihood? lookupTier;
  var lookupReason = '';
  if (inCauses && inAvoids) {
    lookupTier = AeLikelihood.moderate;
    lookupReason = 'Listed in both candidates and causes.';
  } else if (inCauses) {
    lookupTier = AeLikelihood.high;
    lookupReason = 'Commonly caused by this drug.';
  } else if (inAvoids && fromCauses) {
    lookupTier = AeLikelihood.lowerThanCurrent;
    lookupReason =
        'Recommended switch target for ${ae.label.toLowerCase()}.';
  } else if (inAvoids) {
    lookupTier = AeLikelihood.low;
    lookupReason = 'Not typically caused by this drug.';
  }

  // Comparative hint always wins — it's the most useful signal in
  // a switching context.
  if (lookupTier == AeLikelihood.lowerThanCurrent) {
    return PredictedAe(
      ae: ae,
      likelihood: AeLikelihood.lowerThanCurrent,
      reason: lookupReason,
    );
  }

  if (fieldTier == null && lookupTier == null) return null;

  final winner = _takeHigher(fieldTier, lookupTier);
  final String reason;
  if (fieldTier != null && lookupTier != null) {
    reason = winner == fieldTier
        ? 'Drug profile: ${fieldRisk!.jsonValue} risk.'
        : lookupReason;
  } else if (fieldTier != null) {
    reason = 'Drug profile: ${fieldRisk!.jsonValue} risk.';
  } else {
    reason = lookupReason;
  }
  return PredictedAe(ae: ae, likelihood: winner, reason: reason);
}

AeLikelihood _takeHigher(AeLikelihood? a, AeLikelihood? b) {
  if (a == null && b == null) return AeLikelihood.unknown;
  if (a == null) return b!;
  if (b == null) return a;
  return _likelihoodRank(a) >= _likelihoodRank(b) ? a : b;
}

/// Map an AE id to the matching risk field on the Drug object. We only
/// surface a few common AEs through the typed fields; everything else
/// falls through to the reverse-lookup table.
RiskLevel? _pickRiskField(Drug drug, String aeId) {
  switch (aeId) {
    case 'sedation':
      return drug.sedation;
    case 'eps_akathisia':
      return drug.epsRisk;
    case 'hyperprolactinaemia':
      return drug.prolactinRisk;
    case 'qtc_prolongation':
      return drug.qtcRisk;
    case 'weight_gain':
      return drug.metabolicRisk?.score;
    case 'discontinuation_difficult':
      return drug.discontinuationSyndromeRisk?.score;
    default:
      return null;
  }
}

int _likelihoodRank(AeLikelihood l) {
  switch (l) {
    case AeLikelihood.high:
      return 4;
    case AeLikelihood.moderate:
      return 3;
    case AeLikelihood.low:
      return 2;
    case AeLikelihood.lowerThanCurrent:
      return 1;
    case AeLikelihood.unknown:
      return 0;
  }
}

/// Display label for an [AeLikelihood].
String likelihoodLabel(AeLikelihood l) {
  switch (l) {
    case AeLikelihood.high:
      return 'High likelihood';
    case AeLikelihood.moderate:
      return 'Moderate likelihood';
    case AeLikelihood.low:
      return 'Low likelihood';
    case AeLikelihood.lowerThanCurrent:
      return 'Lower than current drug';
    case AeLikelihood.unknown:
      return 'Unknown';
  }
}

/// Semantic colour-token pair for a likelihood. UI maps to AppColors.
({String bg, String text}) likelihoodColorTokens(AeLikelihood l) {
  switch (l) {
    case AeLikelihood.high:
      return (bg: 'danger', text: 'danger');
    case AeLikelihood.moderate:
      return (bg: 'warning', text: 'warning');
    case AeLikelihood.low:
      return (bg: 'accent', text: 'accent');
    case AeLikelihood.lowerThanCurrent:
      return (bg: 'to', text: 'to');
    case AeLikelihood.unknown:
      return (bg: 'border', text: 'muted');
  }
}
